### PR TITLE
fix(backport): checkout from staging branch and ask to update if needed

### LINKED ISSUE #731: update requested staging branch before backporting in `git-node backport`
Currently, if my `staging` branch is behind `upstream` one or not on the `staging` branch at all I will create a bad backport PR

so we should ask do you want to sync `staging` with `upstream` (if it's behind)

and we should checkout from the staging branch

### PR DESCRIPTION
**Pull request by rluvaton**
Fixes #731

this aligns with steps 1-3 in [`doc/contributing/backporting-to-release-lines.md` - How to submit a backport pull request](https://github.com/nodejs/node/blob/main/doc/contributing/backporting-to-release-lines.md#how-to-submit-a-backport-pull-request)

https://github.com/nodejs/node/blob/c2cd74453e7d2794ad81cab63e68371e08bad04f/doc/contributing/backporting-to-release-lines.md?plain=1#L48-L62

### CODE REVIEW

File: lib/backport_session.js [hunk 1]
```diff
@@ -155,12 +155,33 @@ export default class BackportSession extends Session {
...
+          defaultAnswer: true
+        });
+
+      if (shouldCreateStagingBranch) {
+        this.syncBranchWithUpstream(this.stagingBranch);
+      }
+    } else if (!this.isBranchUpToDateWithUpstream(this.stagingBranch)) {
+      const shouldSyncBranch = await cli.prompt(
+        `It seems like your ${this.stagingBranch} is behind the upstream remote` +
+        'do you want to sync it?', { defaultAnswer: false });
```
Comment (RafaelGSS) [inline lib/backport_session.js:171]:
```suggestion
        'do you want to sync it?', { defaultAnswer: true });
```

**Review by RafaelGSS — APPROVED** [review]

File: lib/backport_session.js [hunk 2]
```diff
@@ -61,7 +61,7 @@ export default class BackportSession extends Session {
         cli.stopSpinner(`${file} does not exist in current working tree`,
           cli.SPINNER_STATUS.WARN);
         continue;
-      };
+      }
```
Comment (Trott) [inline lib/backport_session.js:64]: Unrelated style change? (I'm surprised we don't lint these files.)

Comment (Trott) [reply lib/backport_session.js:64]: (Although, yeah, that semicolon shouldn't be there. Maybe leave it and ignore my comment. But uh, future enhancement for someone: Add linting or adjust the current linting to catch that stuff automatically.)

**Comment by rluvaton** [comment]
I totally with you on adding test but there where no test to begin with

I will try to add one...

File: lib/backport_session.js [hunk 3]
```diff
@@ -226,4 +247,51 @@ export default class BackportSession extends Session {
...
+      return false;
+    }
+  }
+
+  syncBranchWithUpstream(branch) {
+    const currentBranch = this.getCurrentBranch();
+
+    runSync('git',
+      [currentBranch !== branch ? 'fetch' : 'pull', 'upstream', `${branch}:${branch}`, '-f']
+    );
```
Comment (aduh95) [inline lib/backport_session.js:295]:
```suggestion
    runSync('git', ['fetch', upstream, branch]);
    if (currentBranch === branch) runSync('git', ['reset', 'FETCH_HEAD', '--hard']);
```
I'm not even sure the `if` condition is necessary, but I guess I'm not using git like everyone else 🤷‍♂️

**Review by aduh95 — CHANGES_REQUESTED** [review]
Please do not hard code `'upstream'`, and take the name of the upstream from the config instead.

File: lib/backport_session.js (hunk 3 above)
Comment (rluvaton) [reply lib/backport_session.js:295]: the fetch here will not update the local branch...

**Comment by rluvaton** [comment]
> Please do not hard code `'upstream'`, and take the name of the upstream from the config instead.

@aduh95 fixed

File: lib/backport_session.js [hunk 4]
```diff
@@ -226,4 +247,56 @@ export default class BackportSession extends Session {
...
+
+  syncBranchWithUpstream(branch) {
+    const currentBranch = this.getCurrentBranch();
+
+    runSync('git',
+      [
+        currentBranch !== branch ? 'fetch' : 'pull',
+        this.upstream,
+        `${branch}:${branch}`,
+        '-f'
```
Comment (aduh95) [inline lib/backport_session.js:298]: Here's what I happen when I try the command locally:

```console
$ git fetch upstream main:main -f
fatal: refusing to fetch into branch 'refs/heads/main' checked out at '…/node'
```

Are you sure it does work?

Comment (rluvaton) [reply lib/backport_session.js:298]: It's weird, it's in the docs of backporting and it's working for me

Are you currently on main branch

File: lib/backport_session.js (hunk 3 above)
Comment (aduh95) [reply lib/backport_session.js]: I don't use local branches, so I guess that's why this looks weird to me.

**Pull request merged**